hedera-node/src/main/java/com/hedera/services/txns/contract/ContractDeleteTransitionLogic.java

```diff
@@ -21,81 +21,98 @@
 	private final TransactionContext txnCtx;
+	private final GlobalDynamicProperties dynamicProperties;
 	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts;
 
 	private final Function<TransactionBody, ResponseCodeEnum> SEMANTIC_CHECK = this::validate;
 
 	@Inject
 	public ContractDeleteTransitionLogic(
-			HederaLedger ledger,
-			LegacyDeleter delegate,
+			AccountStore accountStore,
 			OptionValidator validator,
 			TransactionContext txnCtx,
+			GlobalDynamicProperties dynamicProperties,
 			Supplier<MerkleMap<EntityNum, MerkleAccount>> contracts
 	) {
-		this.ledger = ledger;
-		this.delegate = delegate;
+		this.accountStore = accountStore;
 		this.validator = validator;
 		this.txnCtx = txnCtx;
+		this.dynamicProperties = dynamicProperties;
 		this.contracts = contracts;
 	}
 
-	@FunctionalInterface
-	public interface LegacyDeleter {
-		TransactionRecord perform(TransactionBody txn, Instant consensusTime);
-	}
-
 	@Override
 	public void doStateTransition() {
-		try {
-			final var contractDeleteTxn = txnCtx.accessor().getTxn();
-			final var op = contractDeleteTxn.getContractDeleteInstance();
+		/* --- Translate from gRPC types --- */
+		final var contractDeleteTxn = txnCtx.accessor().getTxn();
+		final var op = contractDeleteTxn.getContractDeleteInstance();
+
+		final var targetId = Id.fromGrpcAccount(asAccount(op.getContractID()));
+		final var target = accountStore.loadContract(targetId);
+
+		final long zeroBalance = 0L;
+		if (target.getBalance() > zeroBalance) {
+			Id beneficiaryId;
+			Account beneficiary;
 
 			if (op.hasTransferAccountID()) {
-				final var receiver = op.getTransferAccountID();
-				if (ledger.exists(receiver) && ledger.isDetached(receiver)) {
-					txnCtx.setStatus(ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
-					return;
-				}
+				beneficiaryId = Id.fromGrpcAccount(op.getTransferAccountID());
+				beneficiary = accountStore.loadAccount(beneficiaryId);
+			} else if (op.hasTransferContractID()) {
+				beneficiaryId = Id.fromGrpcAccount(asAccount(op.getTransferContractID()));
+				beneficiary = accountStore.loadContract(beneficiaryId);
+			} else {
+				//TODO should we always load a beneficiary account?
+				beneficiaryId = Id.fromGrpcAccount(dynamicProperties.fundingAccount());
+				beneficiary = accountStore.loadAccount(beneficiaryId);
 			}
 
-			var legacyRecord = delegate.perform(contractDeleteTxn, txnCtx.consensusTime());
+			//TODO check relevancy for status: OBTAINER_REQUIRED, OBTAINER_DOES_NOT_EXIST
+			validateFalse(beneficiaryId.equals(targetId), OBTAINER_SAME_CONTRACT_ID);
+			validateFalse(beneficiary.isDetached(), ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
 
-			txnCtx.setStatus(legacyRecord.getReceipt().getStatus());
-		} catch (Exception e) {
-			log.warn("Avoidable exception!", e);
-			txnCtx.setStatus(FAIL_INVALID);
+			beneficiary.initBalance(beneficiary.getBalance() + target.getBalance());
+			target.initBalance(zeroBalance);
+
+			accountStore.persistAccount(beneficiary);
```

Comment:
Can we group both of the two persist() calls under  /* --- Persist the changes --- */